java/ql/lib/semmle/code/java/Annotation.qll

```diff
@@ -146,10 +333,54 @@ class AnnotationType extends Interface {
 
   /** Holds if this annotation type is annotated with the meta-annotation `@Inherited`. */
   predicate isInherited() {
-    exists(Annotation ann |
-      ann.getAnnotatedElement() = this and
-      ann.getType().hasQualifiedName("java.lang.annotation", "Inherited")
-    )
+    getADeclaredAnnotation().getType().hasQualifiedName("java.lang.annotation", "Inherited")
+  }
+
+  /** Holds if this annotation type is annotated with the meta-annotation `@Documented`. */
+  predicate isDocumented() {
+    getADeclaredAnnotation().getType().hasQualifiedName("java.lang.annotation", "Documented")
+  }
+
+  /**
+   * Gets the retention policy of this annotation type, that is, the name of one of the
+   * enum constants of `java.lang.annotation.RetentionPolicy`. If no explicit retention
+   * policy is specified the result is `CLASS`.
+   */
+  string getRetentionPolicy() {
+    if getADeclaredAnnotation() instanceof RetentionAnnotation
+    then result = getADeclaredAnnotation().(RetentionAnnotation).getRetentionPolicy()
+    else
+      // If not explicitly specified retention is CLASS
+      result = "CLASS"
+  }
+
+  /**
+   * Holds if the element type is a possible target for this annotation type.
+   * The `elementType` is the name of one of the `java.lang.annotation.ElementType`
+   * enum constants. If no explicit target is specified for this annotation type
+   * it is considered to be applicable to all elements.
+   */
+  // Note: Cannot use a predicate with string as result because annotation type without
+  // explicit @Target can be applied to all targets, requiring to hardcode element types here
+  bindingset[elementType]
+  predicate isATargetType(string elementType) {
+    if getADeclaredAnnotation() instanceof TargetAnnotation
+    then elementType = getADeclaredAnnotation().(TargetAnnotation).getATargetElementType()
+    else
+      // No Target annotation means "applicable to all contexts" since JDK 14, see https://bugs.openjdk.java.net/browse/JDK-8231435
+      // The compiler does not completely implement that, but pretend it did
```

Comment:
Expand on this?

Reply:
It think I was referring to some bugs with module declarations, such as [JDK-8254023](https://bugs.openjdk.java.net/browse/JDK-8254023); I will expand the comment.

However, the behavior for a missing `@Target` annotation changed again in Java 17, now it only includes declaration contexts, but excludes type usage contexts ([JDK-8261610](https://bugs.openjdk.java.net/browse/JDK-8261610)). I will mention this in the comment as well.
I will probably still keep the predicate as `isATargetType` for now but exclude `TYPE_USE` if no `@Target` annotation is present, this seems slightly less error-prone than a `getATargetType()` predicate which is more likely to become outdated as the JDK evolves and new declaration contexts are added.

Or what do you think?

Reply:
How about naming the predicate `mayBeTargetType`, and note in qldoc that the meaning of no-`@Target` has changed over time, with this predicate taking the broadest interpretation?

Reply:
`mayBeTargetType` (or rather <code>mayBe<ins>A</ins>TargetType</code>?) sounds a bit weird to me, but if you want I can rename it nonetheless.
Though extending the documentation sounds like a good idea, but I think I will rather mention "latest Java version" or similar, since "broadest" might sound like applicable in all contexts (including type contexts), which is no longer the case since Java 17.

Reply:
`getA...` would have to be a predicate with result then, right? That won't be possible with the current implementation.
My main concerns with `mayBeTargetType` is that personally I find that "may be" sounds a bit weird (and too close to "maybe", especially as predicate / method prefix), but this is probably rather subjective; and that this name expresses some uncertainty (I think), even though the predicate is fairly reliable (unless you provide it with strings which are not `ElementType` values).